browse/controllers/archive_page/__init__.py

```diff
@@ -0,0 +1,140 @@
+
+    subsumed_by = ARCHIVES_SUBSUMED.get(archive_id, None)
+    if subsumed_by:
+        data["subsumed_id"] = archive_id
+        data["subsumed_category"] = CATEGORIES.get(archive_id, {})
+        data["subsumed_by"] = subsumed_by
+        subsuming_category = CATEGORIES.get(subsumed_by, {})
+        data["subsuming_category"] = subsuming_category
+        archive_id = subsuming_category.get("in_archive", None)
+        archive = ARCHIVES.get(archive_id, None)
+
+    years = years_operating(archive)
+
+    data["archive_id"] = archive_id
+    data["archive"] = archive
+    data["list_form"] = ByMonthForm(archive_id, archive, years)
+    data["catchup_form"] = CatchupForm(archive_id, archive, years)
+    data["stats_by_year"] = stats_by_year(archive_id, archive, years)
+    data["category_list"] = category_list(archive_id)
+
+    data["template"] = "archive/single_archive.html"
+    return data, status.HTTP_200_OK, response_headers  # type: ignore
+
+
+def archive_index(archive_id: str, status: int) -> Response:
+    """Landing page for when there is no archive specified."""
+    data: Dict[str, Any] = {}
+    data["bad_archive"] = archive_id
+
+    archives = [
+        (id, ARCHIVES[id]["name"])
+        for id in ARCHIVES.keys()
+        if id not in ARCHIVES_SUBSUMED and not id.startswith("test")
+    ]
+    archives.sort(key=lambda tpl: tpl[0])
+    data["archives"] = archives
+
+    defunct = [
+        (id, ARCHIVES[id]["name"], ARCHIVES_SUBSUMED.get(id, ""))
+        for id in ARCHIVES.keys()
+        if "end_date" in ARCHIVES[id]
+    ]
+    defunct.sort(key=lambda tpl: tpl[0])
+    data["defunct"] = defunct
+    
+    data["template"] = "archive/archive_list_all.html"
+    return data, status, {}  # type: ignore
+
+
+def subsumed_msg(archive: Dict[str, str], subsumed_by: str) -> Dict[str, str]:
+    """Adds information about subsuming categories and archives."""
+    sb = CATEGORIES.get(subsumed_by, {"name": "unknown category"})
+    sa = ARCHIVES.get(sb.get("in_archive", None), {"name": "unknown archive"})
+
+    return {"subsumed_by_cat": sb, "subsumed_by_arch": sa}
+
+
+def years_operating(archive: Dict[str, Any]) -> List[int]:
+    """Returns list of years operating in desc order. ex [1993,1992,1991]."""
+    if (
+        not archive
+        or "start_date" not in archive
+        or not isinstance(archive["start_date"], date)
+    ):
+        return []
+    start = archive["start_date"].year
+    end = archive.get("end_date", None) or date.today().year
+    return list(reversed(range(start, end + 1)))
+
+
+def stats_by_year(
+    archive_id: str, archive: Dict[str, Any], years: List[int]
+) -> List[Tuple[Any, str]]:
+    """Returns links to year pages."""
+    if not archive or not archive_id or not years:
+        return [("bogusURL", "NODATA")]
+    else:
+        return [(_year_stats_link(archive_id, i), str(i)) for i in years]
+
+
+def _year_stats_link(archive_id: str, num: int) -> Any:
+    return url_for(
```

Comment:
Does `url_for` return a string, or can it be treated as such? Might reduce many of the `Any`s found in the surrounding code. Mabe we could make our own `url_for` wrapper in arxiv-base to make it more clear if not. Sorry for all the nitpicks on this, and I know it is usually flask and friends at fault, but seeing `Any` always makes me wonder what is going on.

Reply:
I'm thinking that we could use more @no_type_check and file or block  # type: ignore ( https://github.com/python/mypy/issues/626 ).
In arXiv specific non-flask business logic code mypy might still be useful.